csharp/ql/lib/semmle/code/csharp/dataflow/internal/ContentDataFlow.qll

```diff
@@ -0,0 +1,432 @@
+        state instanceof StoreState or
+        state instanceof ReadState
+      )
+    }
+
+    final override predicate isAdditionalFlowStep(
+      Node node1, DF::FlowState state1, Node node2, DF::FlowState state2
+    ) {
+      storeStep(node1, state1, _, node2, state2, this) or
+      readStep(node1, state1, _, node2, state2, this) or
+      additionalStep(node1, state1, node2, state2, this)
+    }
+
+    final override predicate isBarrier(Node node) { c.isBarrier(node) }
+
+    final override FlowFeature getAFeature() { result = c.getAFeature() }
+
+    final override predicate includeHiddenNodes() { any() }
+  }
+
+  private newtype TAccessPath =
+    TAccessPathNil() or
+    TAccessPathCons(ContentSet head, AccessPath tail) {
+      nodeReachesStore(_, _, _, _, head, _, tail)
+      or
+      nodeReachesRead(_, _, _, _, head, tail, _)
+    }
+
+  /** An access path. */
+  class AccessPath extends TAccessPath {
+    /** Gets the head of this access path, if any. */
+    ContentSet getHead() { this = TAccessPathCons(result, _) }
+
+    /** Gets the tail of this access path, if any. */
+    AccessPath getTail() { this = TAccessPathCons(_, result) }
+
+    /**
+     * Gets a textual representation of this access path.
+     *
+     * Elements are dot-separated, and the head of the stack is
+     * rendered first.
+     */
+    string toString() {
+      this = TAccessPathNil() and
+      result = ""
+      or
+      exists(ContentSet head, AccessPath tail |
+        this = TAccessPathCons(head, tail) and
+        result = head + "." + tail
+      )
+    }
+  }
+
+  pragma[nomagic]
+  private predicate pathNodeSuccEntry(DF::PathNode node) {
+    node.getConfiguration() instanceof Configuration and
+    (
+      any(ConfigurationAdapter config).isSource(node.getNode(), node.getState())
+      or
+      DF::PathGraph::subpaths(_, _, _, node)
+      or
+      DF::PathGraph::subpaths(_, node, _, _)
+      or
+      storeStep(_, _, _, node.getNode(), node.getState(), node.getConfiguration())
+      or
+      readStep(_, _, _, node.getNode(), node.getState(), node.getConfiguration())
+    )
+  }
+
+  pragma[nomagic]
+  private predicate pathNodeSuccExit(DF::PathNode node) {
+    node.getConfiguration() instanceof Configuration and
+    (
+      any(ConfigurationAdapter config).isSink(node.getNode(), node.getState())
+      or
+      DF::PathGraph::subpaths(node, _, _, _)
+      or
+      DF::PathGraph::subpaths(_, _, node, _)
+      or
+      storeStep(node.getNode(), node.getState(), _, _, _, node.getConfiguration())
+      or
+      readStep(node.getNode(), node.getState(), _, _, _, node.getConfiguration())
+    )
+  }
+
+  pragma[nomagic]
+  private predicate pathNodeSuccRec(DF::PathNode pred, DF::PathNode succ) {
+    pathNodeSuccEntry(pred) and
+    succ = pred
+    or
+    exists(DF::PathNode mid |
+      pathNodeSuccRec(pred, mid) and
+      succ = mid.getASuccessor() and
+      not DF::PathGraph::subpaths(mid, succ, _, _) and
```

Comment:
Don't you need `not DF::PathGraph::subpaths(mid, _, _, succ) and` as well here?

Reply:
Both are actually redundant, as they are implied by `pathNodeSuccExit(mid)`. But it made me realize that there is one case that doesn't work: Flow into a parameter (and not back out) for which there is also a subpath.